File: src/algorithms/barycentric_triangle.rs
use crate::common::{DVector, Point, Triangle};
use rayon::prelude::*;

fn compute_weights(triangle: Triangle, p: Point) -> (f64, f64, f64) {
    let ab = triangle.b - triangle.a;
    let ac = triangle.c - triangle.a;
    let ap = p - triangle.a;

    let Aa = ab.cross(ap).abs();
    let Ac = ac.cross(ap).abs();
    let A = ab.cross(ac).abs();

    let inv_denom = 1.0 / A;
    let w = inv_denom * Aa;
    let v = inv_denom * Ac;
    let u = 1.0 - v - w;

    return (u, v, w);
}

fn barycentric_triangle_weights(
    points: &[&[f64]],
    face_indices: &[i64],
    faces: &[&[i64; 3]],
    vertices: &[&[f64; 2]],
) -> Vec<[f64; 3]> {
    points
        .par_iter()
        .zip(face_indices)
        .map(|(point, face_index)| {
            if face_index == &-1 {
                return [0., 0., 0.];
            }
            todo!()
        })
        .collect::<Vec<_>>();
    todo!()
}

File: numba_celltree/algorithms/barycentric_triangle.py
"""
This is a straightforward implementation of barycentric interpolation for
triangles.

For barycentric interpolation on N-sided convex polygons, see
barycentric_wachspress.py.
"""

import numba as nb
import numpy as np

from numba_celltree.constants import (
    PARALLEL,
    FloatArray,
    FloatDType,
    IntArray,
    Triangle,
)
from numba_celltree.geometry_utils import (
    Point,
    as_point,
    as_triangle,
    cross_product,
    to_vector,
)


@nb.njit(inline="always")
def compute_weights(triangle: Triangle, p: Point, weights: FloatArray):
    ab = to_vector(triangle.a, triangle.b)
    ac = to_vector(triangle.a, triangle.c)
    ap = to_vector(triangle.a, p)
    Aa = abs(cross_product(ab, ap))
    Ac = abs(cross_product(ac, ap))
    A = abs(cross_product(ab, ac))
    inv_denom = 1.0 / A
    w = inv_denom * Aa
    v = inv_denom * Ac
    u = 1.0 - v - w
    weights[0] = u
    weights[1] = v
    weights[2] = w
    return


@nb.njit(parallel=PARALLEL, cache=True)
def barycentric_triangle_weights(
    points: FloatArray,
    face_indices: IntArray,
    faces: IntArray,
    vertices: FloatArray,
) -> FloatArray:
    n_points = len(points)
    weights = np.zeros((n_points, 3), dtype=FloatDType)
    for i in nb.prange(n_points):
        face_index = face_indices[i]
        if face_index == -1:
            continue
        face = faces[face_index]
        triangle = as_triangle(vertices, face)
        point = as_point(points[i])
        compute_weights(triangle, point, weights[i])
    return weights
